### PR TITLE
Added get_versions method in NodeJS api with the basic test.

### LINKED ISSUE #23430: [Good first issue] Port Core.get_versions() method to JS API
### Context

JS API is a young branch of other OpenVINO bindings.
The bindings essentially map the functions or methods exposed by the API in the original language to equivalent constructs in the target language.
Original language in this case it's C++ and JS is the target language.
We propose to start learning OpenVINO from the implementation simple feature for JS API.

**Port **Core.get_versions()** method to JS API**

In C++ API, Core has [get_versions](https://docs.openvino.ai/2024/api/c_cpp_api/classov_1_1_core.html#_CPPv4NK2ov4Core12get_versionsERKNSt6stringE) method.
It would be great to have this method in JS API. You need to implement Node.js bindings for this method.
As examples of implementation, please, review Example PRs section below.

You can learn:
- How to compile OpenVINO from sources
- How Node.js bindings works
- Write some TypeScript code
- Be prepared for the next feature request in big opensource product

### What needs to be done?

- Add interface implementation to addon.ts
- Write js test for this method
- Add method definition into bindings (core_wrap.hpp)
- Implement this method in bindings (core_wrap.cpp)

### Example Pull Requests

- https://github.com/openvinotoolkit/openvino/pull/22100
- https://github.com/openvinotoolkit/openvino/pull/23338

### Resources
- [JS API Reference](https://docs.openvino.ai/2024/api/nodejs_api/addon.html)
- [Node.js bindings examples of usage](https://github.com/nodejs/node-addon-examples)
- [Contribution guide - start here!](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [Blog post on contributing to OpenVINO](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [User documentation](https://docs.openvino.ai/)

### Contact points

@vishniakov-nikolai @almilosz

### Ticket

134816

### PR DESCRIPTION
**Pull request by adismort14**
This Fixes #23430 .

Details:

1. Added get_versions() method definition into bindings (core_wrap.hpp)
2. Implemented this method in bindings (core_wrap.cpp)
3. Add interface implementation to addon.ts
4. Added basic js test for this method in `src/bindings/js/node/tests/basic.test.js`

CC: @vishniakov-nikolai

### CODE REVIEW

File: src/bindings/js/node/lib/addon.ts [hunk 1]
```diff
@@ -38,6 +38,7 @@ interface Core {
   readModelSync(modelBuffer: Uint8Array, weightsBuffer?: Uint8Array): Model;
   importModelSync(modelStream: Buffer, device: string): CompiledModel;
   getAvailableDevices(): string[];
+  getVersions():{ deviceName: string; version: string }[];
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/lib/addon.ts:41]: Misalignment with bindings signature

File: src/bindings/js/node/src/core_wrap.cpp [hunk 2]
```diff
@@ -232,6 +233,25 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
     return js_devices;
 }
 
+Napi::Value CoreWrap::get_versions(const Napi::CallbackInfo& info) {
+    const auto& devices_map = _core.get_versions(info[0].As<Napi::String>().Utf8Value());
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/src/core_wrap.cpp:237]: What if not string type would be passed or args would absent?

File: src/bindings/js/node/src/core_wrap.cpp [hunk 3]
```diff
@@ -232,6 +233,25 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
...
+    Napi::Array js_devices = Napi::Array::New(info.Env(), devices_map.size());
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/src/core_wrap.cpp:238]: Why array should have size = passed string length?

File: src/bindings/js/node/src/core_wrap.cpp [hunk 4]
```diff
@@ -232,6 +233,25 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
...
+
+    size_t i = 0;
+    for (const auto& dev : devices_map) {
+        Napi::Object js_device = Napi::Object::New(info.Env());
+        js_device.Set("deviceName", Napi::String::New(info.Env(), dev.first));
+        std::string versionStr = dev.second.buildNumber;
+        versionStr += " ";
+        versionStr += dev.second.description;
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/src/core_wrap.cpp:246]: I propose to use structure, do not concatenate parameters in string

File: src/bindings/js/node/tests/basic.test.js [hunk 5]
```diff
@@ -20,6 +20,13 @@ it('Core.getAvailableDevices()', () => {
     assert.ok(devices.includes('CPU'));
 });
 
+it('Core.getVersions()', () => {    
+  const devices = core.getAvailableDevices();
+  const deviceVersion = core.getVersions(devices[0]);
+
+  assert.ok(deviceVersion.size()>=1);
+});
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/tests/basic.test.js:28]: Tests have to check behavior in the cases when incorrect parameters passed, parameters are absent and that result has expected format.

**Review by vishniakov-nikolai — CHANGES_REQUESTED** [review]
Please, pay attention on comments. If you need more information, feel free to ask.

File: src/bindings/js/node/src/core_wrap.cpp (hunk 3 above)
Comment (adismort14) [reply src/bindings/js/node/src/core_wrap.cpp:238]: `_core.get_versions` returns a map containing the device and the deviceVersion. Thus the length of js_devices array should be the same as that of devices_map.
I have referred to this doc for the same. 
```std::map<std::string, ov::Version> ov::Core::get_versions(const std::string &device_name) const
Returns device plugins version information. Device name can be complex and identify multiple devices at once like HETERO:CPU,GPU; in this case, std::map contains multiple entries, each per device.

Parameters:
device_name – Device name to identify a plugin.

Returns:
A vector of versions.
```

File: src/bindings/js/node/lib/addon.ts (hunk 1 above)
Comment (adismort14) [reply src/bindings/js/node/lib/addon.ts]: Fixed this in the latest commit. Kindly review it.

File: src/bindings/js/node/src/core_wrap.cpp (hunk 2 above)
Comment (adismort14) [reply src/bindings/js/node/src/core_wrap.cpp]: I have added the required checks for this.

**Comment by adismort14** [comment]
Here are the updates from my end post the review:

1. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1539573589) Fixed the binding signatures.
2. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1539575566) Added the check for no args passed and non string convertible arguments passed.
3. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1539604579) Modified the code to contain the parameters in an object of the form:
```js
{
    deviceName: String,
    version: String,
    versionDescription: String
}
```
4. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1539625384) Added more tests to test the unexpected behaviour. I have a question regarding this: How do I run the tests locally to check whether the tests I have written work as expected and not have to discover them during the github workflow checks? Could you share any doc or share some pointers for the same?

@vishniakov-nikolai  Kindly review them and let me know if any changes are required.
Thanks!

File: src/bindings/js/node/lib/addon.ts [hunk 6]
```diff
@@ -38,6 +38,7 @@ interface Core {
   readModelSync(modelBuffer: Uint8Array, weightsBuffer?: Uint8Array): Model;
   importModelSync(modelStream: Buffer, device: string): CompiledModel;
   getAvailableDevices(): string[];
+  getVersions(deviceName: string):{ deviceName: string, version: string, versionDescription: string }[];
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/lib/addon.ts:41]: I propose to follow original format: 
```suggestion
  getVersions(deviceName: string): { 
    [deviceName: string]: { 
      buildNumber: string,
      description: string,
    },
  };
```

File: src/bindings/js/node/src/core_wrap.cpp [hunk 7]
```diff
@@ -232,6 +233,40 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
     return js_devices;
 }
 
+Napi::Value CoreWrap::get_versions(const Napi::CallbackInfo& info) {
+    // If no arguments are passed into the method.
+    if(info.Length()==0) {
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/src/core_wrap.cpp:238]: Please, make sure that your code formatted using openvino [clang format](https://github.com/openvinotoolkit/openvino/blob/master/src/.clang-format).

File: src/bindings/js/node/src/core_wrap.cpp (hunk 3 above)
Comment (vishniakov-nikolai) [reply src/bindings/js/node/src/core_wrap.cpp:249]: My bad, I wrongly interpreted map type.

File: src/bindings/js/node/src/core_wrap.cpp [hunk 8]
```diff
@@ -232,6 +233,40 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
...
+
+    size_t i = 0;
+    for (const auto& dev : devices_map) {
+        Napi::Object js_device = Napi::Object::New(info.Env());
+        // js_device Object is of the form:
+        // {
+        //   deviceName: String,
+        //   version: String,
+        //   versionDescription: String
+        // }
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/src/core_wrap.cpp:259]: Comment like this is extra. Code should be selfdocumented.

File: src/bindings/js/node/lib/addon.ts (hunk 6 above)
Comment (vishniakov-nikolai) [reply src/bindings/js/node/lib/addon.ts:41]: Also, please check that your IDE is using and following project eslint rules

**Comment by adismort14** [comment]
Hey!
I have modified the code based on your review.

1. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1543177538) I have modified the binding signature according to this as well as the get_versions function code in `core_wrap.cpp`.
2. [Review Comment](https://github.com/openvinotoolkit/openvino/pull/23648#discussion_r1543183489) Formatted the code using clang format.
3. I have also corrected my tests based on the modifications. I have run them locally and they are passing. Removed all the unwanted comments and eslint changes.

@vishniakov-nikolai Please do review the changes when you get the time. :smile:

File: src/bindings/js/node/src/core_wrap.cpp [hunk 9]
```diff
@@ -232,6 +233,31 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
     return js_devices;
 }
 
+Napi::Value CoreWrap::get_versions(const Napi::CallbackInfo& info) {
+    if (info.Length() == 0) {
+        reportError(info.Env(), "No argument provided in the getVersions() method call.");
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/src/core_wrap.cpp:238]: Include useful info into error message
```suggestion
        reportError(info.Env(), "get_versions expects 1 argument of string type");
```

File: src/bindings/js/node/src/core_wrap.cpp [hunk 10]
```diff
@@ -232,6 +233,31 @@ Napi::Value CoreWrap::get_available_devices(const Napi::CallbackInfo& info) {
...
+        return info.Env().Undefined();
+    }
+    auto device_arg = info[0];
+    if (!device_arg.IsString()) {
+        reportError(info.Env(), "The argument in getVersions() method must be a string or convertible to a string.");
+        return info.Env().Undefined();
+    }
+    const auto& devices_map = _core.get_versions(device_arg.ToString());
+    Napi::Object versionsObject = Napi::Object::New(info.Env());
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/src/core_wrap.cpp:247]: Better to use snake case notation at c++ side
```suggestion
    Napi::Object versions_object = Napi::Object::New(info.Env());
```

File: src/bindings/js/node/tests/basic.test.js [hunk 11]
```diff
@@ -20,6 +20,31 @@ it('Core.getAvailableDevices()', () => {
     assert.ok(devices.includes('CPU'));
 });
 
+describe('Core.getVersions()', () => {
+  const devices = core.getAvailableDevices();
+
+  it('getVersions(validDeviceName: string)', () => {    
+    const deviceVersion = core.getVersions(devices[0]);
+    assert.ok(Object.keys(deviceVersion).length >= 1);
+  });
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/tests/basic.test.js:29]: I don't see any purpose to check it in this way. 
My proposals:
- check that `getVersions` returns object
- check that object contains property in specified format

**Review by vishniakov-nikolai — COMMENTED** [review]
Almost ready, minor comments

File: src/bindings/js/node/tests/basic.test.js [hunk 12]
```diff
@@ -20,6 +20,34 @@ it('Core.getAvailableDevices()', () => {
...
+  it('getVersions() throws if no arguments are passed into the function', () => {
+    assert.throws(
+      () => core.getVersions(),
+      {message: 'getVersions() method expects 1 argument of string type.'}
+    );
+  });
+
+  it('getVersions() throws if non string coercable arguments are passed into the function', () => {
+    assert.throws(
+      () => core.getVersions({deviceName:devices[0]}),
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/tests/basic.test.js:43]:
```suggestion
      () => core.getVersions({ deviceName: devices[0] }),
```

**Comment by vishniakov-nikolai** [comment]
@adismort14 please, merge latest master to avoid conflicts

File: src/bindings/js/node/tests/basic.test.js [hunk 13]
```diff
@@ -20,6 +20,34 @@ it('Core.getAvailableDevices()', () => {
     assert.ok(devices.includes('CPU'));
 });
 
+describe('Core.getVersions()', () => {
+  const devices = core.getAvailableDevices();
+
+  it('getVersions(validDeviceName: string)', () => {    
+    const deviceVersion = core.getVersions(devices[0]);
+    assert.strictEqual(typeof deviceVersion, 'object');
+    assert.strictEqual(typeof deviceVersion.CPU, 'object');
```
[the remark this answers is not in the document]
Comment (adismort14) [reply src/bindings/js/node/tests/basic.test.js]: I have hardcoded the value to 'CPU' in this [commit](https://github.com/openvinotoolkit/openvino/pull/23648/commits/ad2dbcfbb189c7331f32f6832a3e48fd3b7c2a09).

**Comment by adismort14** [comment]
Do share your review based on the latest commit. Have merged and resolved the merge conflicts. @vishniakov-nikolai

**Pull request merged**